src/NuGet.Core/NuGet.Protocol.Core.v2/DownloadResourceV2.cs

```diff
@@ -38,34 +40,58 @@ public DownloadResourceV2(V2Resource resource)
                 throw new ArgumentNullException(nameof(identity));
             }
 
+            string uri = string.Empty;
+
             return Task.Run(() =>
             {
                 token.ThrowIfCancellationRequested();
 
-                try
+                for (int i = 0; i < 3; i++)
                 {
-                    var sourcePackage = identity as SourcePackageDependencyInfo;
-                    var repository = V2Client as DataServicePackageRepository;
+                    try
+                    {
+                        var sourcePackage = identity as SourcePackageDependencyInfo;
+                        var repository = V2Client as DataServicePackageRepository;
+
+                        bool isFromUri = repository != null
+                            && sourcePackage?.PackageHash != null
+                            && sourcePackage?.DownloadUri != null;
+
+                        if (isFromUri)
+                        {
+                            uri = sourcePackage?.DownloadUri.AbsoluteUri;
+                        }
+                        else
+                        {
+                            uri = (repository as DataServicePackageRepository)?.Source ?? "Folder";
```

Comment:
this would be more clear if uri was named, displayPath. Uri makes it sound like this path will be used for the download.